runtime/j9vm/java11vmi.c

```diff
@@ -710,18 +730,80 @@ allowReadAccessToModule(J9VMThread * currentThread, J9Module * fromModule, J9Mod
  * @return If successful, returns a java.lang.reflect.Module object. Otherwise, returns NULL.
  */
 jobject JNICALL
+#if JAVA_SPEC_VERSION >= 15
+JVM_DefineModule(JNIEnv * env, jobject module, jboolean isOpen, jstring version, jstring location, jobjectArray packageArray)
+#else
 JVM_DefineModule(JNIEnv * env, jobject module, jboolean isOpen, jstring version, jstring location, const char* const* packages, jsize numPackages)
+#endif /* JAVA_SPEC_VERSION >= 15 */
 {
 	J9VMThread * const currentThread = (J9VMThread*)env;
 	J9JavaVM * vm = currentThread->javaVM;
 	J9InternalVMFunctions const * const vmFuncs = vm->internalVMFunctions;
+#if JAVA_SPEC_VERSION >= 15
+	BOOLEAN oom = FALSE;
+	BOOLEAN nullPackage = FALSE;
+	jsize numPackages = (NULL == packageArray) ? 0 : (*env)->GetArrayLength(env, packageArray);
+	PORT_ACCESS_FROM_ENV(env);
+	UDATA packagesNumBytes = sizeof(char*) * numPackages;
+	const char* * packages = (const char* *)j9mem_allocate_memory(packagesNumBytes, OMRMEM_CATEGORY_VM);
+	if ((*env)->ExceptionCheck(env)) {
+		goto cleanup;
+	}
+
+	if (NULL != packages) {
+		jsize pkgIndex = 0;
+		memset((void *)packages, 0, packagesNumBytes);
+		for (pkgIndex = 0; pkgIndex < numPackages; pkgIndex++) {
+			jobject packageObj = (*env)->GetObjectArrayElement(env, packageArray, pkgIndex);
+			if ((*env)->ExceptionCheck(env)) {
+				Assert_SC_unreachable();
+				goto cleanup;
+			}
+			if (NULL != packageObj) {
+				jboolean isCopy = JNI_FALSE;
+				const char* packageName = (*env)->GetStringUTFChars(env, packageObj, &isCopy);
+				(*env)->DeleteLocalRef(env, packageObj);
+
+				if (NULL == packageName) {
+					oom = TRUE;
+					break;
+				}
+				if (JNI_FALSE == isCopy) {
+					Assert_SC_unreachable();
+					goto cleanup;
+				}
+				convertPackageName((char*)packageName);
```

Comment:
It would be more efficient to get vm access earlier and do this with VM calls rather than JNI.  It would also remove the need to assert on the `isCopy` above by using the `copyStringToUTF8Helper` helper

Reply:
Updated to use copyStringToUTF8Helper().